### PR TITLE
azurerm_express_route_circuit_peering -  support for bandwidth_in_gbps, express_route_port_id

### PR DESCRIPTION
**Pull request by xuzhang3**

### CODE REVIEW

File: website/docs/r/express_route_circuit.html.markdown [hunk 1]
```diff
@@ -47,17 +47,27 @@ The following arguments are supported:
...
 
-* `bandwidth_in_mbps` - (Required) The bandwidth in Mbps of the circuit being created.
+* `bandwidth_in_mbps` - (Optional) The bandwidth in Mbps of the circuit being created on the Service Provider.
 
 ~> **NOTE:** Once you increase your bandwidth, you will not be able to decrease it to its previous value.
 
-* `sku` - (Required) A `sku` block for the ExpressRoute circuit as documented below.
+~> **NOTE:** The `service_provider_name`, the `peering_location` and the `bandwidth_in_mbps` should be set together and they conflict with `express_route_port_id` and `bandwidth_in_gbps`.
+
+* `allow_classic_operations` - (Optional) Allow the circuit to interact with classic (RDFE) resources.
```
Comment (magodo) [inline website/docs/r/express_route_circuit.html.markdown:62]:
```suggestion
* `allow_classic_operations` - (Optional) Allow the circuit to interact with classic (RDFE) resources. Defaults to `false`.
```

File: website/docs/r/express_route_circuit.html.markdown [hunk 2]
```diff
@@ -47,17 +47,27 @@ The following arguments are supported:
...
+
+~> **NOTE:** The `allow_classic_operations` cannot be set while using the Express Route Port.
```
Comment (magodo) [inline website/docs/r/express_route_circuit.html.markdown:64]: Can we add this constraint back in the schema definition?

File: website/docs/r/express_route_circuit.html.markdown [hunk 3]
```diff
@@ -47,17 +47,27 @@ The following arguments are supported:
...
 
-* `allow_classic_operations` - (Optional) Allow the circuit to interact with classic (RDFE) resources. The default value is `false`.
+* `express_route_port_id` - (Optional) The ID of the Express Route Port which this circuit should be based on.
```
Comment (magodo) [inline website/docs/r/express_route_circuit.html.markdown:66]:
```suggestion
* `express_route_port_id` - (Optional) The ID of the Express Route Port which this Express Route Circuit is based on.
```

**Review by magodo — CHANGES_REQUESTED** [review]
@xuzhang3 Thank you for the PR! I've looked through and left some comments.

File: azurerm/internal/services/network/express_route_circuit_resource.go [hunk 4]
```diff
@@ -283,6 +309,16 @@ func resourceExpressRouteCircuitRead(d *pluginsdk.ResourceData, meta interface{}
 		}
 	}
 
+	if resp.ExpressRoutePort != nil {
+		d.Set("bandwidth_in_gbps", resp.BandwidthInGbps)
+
+		portID, err := parse.ExpressRoutePortID(*resp.ExpressRoutePort.ID)
```
Comment (magodo) [inline azurerm/internal/services/network/express_route_circuit_resource.go:315]: Can we nil check the `resp.ExpressRoutePort.ID` here?

File: website/docs/r/express_route_circuit.html.markdown [hunk 5]
```diff
@@ -47,17 +47,25 @@ The following arguments are supported:
 
 * `location` - (Required) Specifies the supported Azure location where the resource exists. Changing this forces a new resource to be created.
 
-* `service_provider_name` - (Required) The name of the ExpressRoute Service Provider.
+* `sku` - (Required) A `sku` block for the ExpressRoute circuit as documented below.
+
+* `service_provider_name` - (Optional) The name of the ExpressRoute Service Provider. Any change to the configuration will recreate the resource.
 
-* `peering_location` - (Required) The name of the peering location and **not** the Azure resource location.
+* `peering_location` - (Optional) The name of the peering location and **not** the Azure resource location. Any change to the configuration will recreate the resource.
```
Comment (magodo) [inline website/docs/r/express_route_circuit.html.markdown:54]:
```suggestion
* `peering_location` - (Optional) The name of the peering location and **not** the Azure resource location. Changing this forces a new resource to be created.
```

File: website/docs/r/express_route_circuit.html.markdown [hunk 6]
```diff
@@ -47,17 +47,25 @@ The following arguments are supported:
 
 * `location` - (Required) Specifies the supported Azure location where the resource exists. Changing this forces a new resource to be created.
 
-* `service_provider_name` - (Required) The name of the ExpressRoute Service Provider.
+* `sku` - (Required) A `sku` block for the ExpressRoute circuit as documented below.
+
+* `service_provider_name` - (Optional) The name of the ExpressRoute Service Provider. Any change to the configuration will recreate the resource.
```
Comment (magodo) [inline website/docs/r/express_route_circuit.html.markdown:52]:
```suggestion
* `service_provider_name` - (Optional) The name of the ExpressRoute Service Provider. Changing this forces a new resource to be created.
```

**Review by magodo — CHANGES_REQUESTED** [review]
Thanks for the update!
I have left some additional mostly minor comments that once addressed this should be good to merge 👍

**Comment by magodo** [comment]
@xuzhang3 overall it LGTM! Despite, the test has one small error, which looks like wants to adding some sequential in the acctest. If we can fix that, it should be good to go!

image

**Comment by xuzhang3** [comment]
> @xuzhang3 overall it LGTM! Despite, the test has one small error, which looks like wants to adding some sequential in the acctest. If we can fix that, it should be good to go!
> 
> image

Every subscription can create no more than 10 circuits by default. https://docs.microsoft.com/en-us/azure/expressroute/expressroute-faqs#can-i-have-more-than-one-expressroute-circuit-in-my-subscription

**Comment by magodo** [comment]
@xuzhang3 So can we reflect that in the acctest?

**Comment by xuzhang3** [comment]
Fixed

> @xuzhang3 So can we reflect that in the acctest?

Fixed

**Comment by magodo** [comment]
LGTM now, thank you!
image

**Review by katbyte — APPROVED** [review]
LGTM 🍰

**Pull request merged**